company-lsp.el

```diff
@@ -305,22 +328,18 @@ See the documentation of `company-backends' for COMMAND and ARG."
              (lsp--capability "completionProvider")
              (or (company-lsp--completion-prefix) 'stop)))
     (candidates
-     (cons :async
-           #'(lambda (callback)
-               (let ((cache (company-lsp--cache-get arg)))
-                 (if cache
-                     (funcall callback (plist-get cache :candidates))
-                   (let ((req (lsp--make-request "textDocument/completion"
-                                                 (lsp--text-document-position-params))))
-                     (setq company-lsp--line-backup
-                           (buffer-substring (line-beginning-position) (line-end-position)))
-                     (if company-lsp-async
-                         (lsp--send-request-async req
-                                                  (lambda (resp)
-                                                    (company-lsp--on-completion resp arg callback)))
-                       (company-lsp--on-completion (lsp--send-request req)
-                                                   arg
-                                                   callback))))))))
+     ;; If the completion items in the response have textEdit action populated,
+     ;; we'll apply them in `company-lsp--post-completion'. However, textEdit
+     ;; actions only apply to the pre-completion content. We backup the current
+     ;; line and restore it after company completion is done, so the content
+     ;; is restored and textEdit actions can be applied.
```

Comment:
Which pre-completion context is this referring to? No prefix typed at all? With prefix, but without its last character? Or either, depending on the situation?

Reply:
With prefix, but without the completed suffix. Essentially we want to undo the completion by company.

Reply:
Hmm, yeah, that's suboptimal. Not sure what to suggest. Company could provide more info to `post-completion`, though.